File: Tests/MemoTextLogicCheck.swift
import Foundation

@main
enum MemoTextLogicCheck {
    static func main() throws {
        let continued = MemoTextLogic.listEdit(
            in: "1. first" as NSString,
            selectedRange: NSRange(location: 8, length: 0)
        )
        assert(continued == MemoTextEdit(range: NSRange(location: 8, length: 0), replacement: "\n2. "))

        let ended = MemoTextLogic.listEdit(
            in: "1. first\n2. " as NSString,
            selectedRange: NSRange(location: 12, length: 0)
        )
        assert(ended == MemoTextEdit(range: NSRange(location: 9, length: 3), replacement: ""))

        let bullet = MemoTextLogic.listEdit(
            in: "- first" as NSString,
            selectedRange: NSRange(location: 7, length: 0)
        )
        assert(bullet == MemoTextEdit(range: NSRange(location: 7, length: 0), replacement: "\n- "))

        let resumedNumber = MemoTextLogic.listEdit(
            in: "1. first\n\t- detail\n\t- " as NSString,
            selectedRange: NSRange(location: 22, length: 0)
        )
        assert(resumedNumber == MemoTextEdit(range: NSRange(location: 19, length: 3), replacement: "2. "))

        let indentedCaret = MemoTextLogic.indentationEdit(
            in: "hello" as NSString,
            selectedRange: NSRange(location: 2, length: 0),
            outdent: false
        )
        assert(indentedCaret == MemoTextEdit(
            range: NSRange(location: 0, length: 5),
            replacement: "\thello",
            selection: NSRange(location: 3, length: 0)
        ))

        let indentedLines = MemoTextLogic.indentationEdit(
            in: "one\ntwo\nthree" as NSString,
            selectedRange: NSRange(location: 1, length: 6),
            outdent: false
        )
        assert(indentedLines == MemoTextEdit(
            range: NSRange(location: 0, length: 8),
            replacement: "\tone\n\ttwo\n",
            selection: NSRange(location: 2, length: 7)
        ))

        let outdentedLines = MemoTextLogic.indentationEdit(
            in: "\tone\n\ttwo" as NSString,
            selectedRange: NSRange(location: 0, length: 9),
            outdent: true
        )
        assert(outdentedLines == MemoTextEdit(
            range: NSRange(location: 0, length: 9),
            replacement: "one\ntwo",
            selection: NSRange(location: 0, length: 7)
        ))

        let expression = try MemoTextLogic.searchExpression(
            pattern: #"item ([0-9]+)"#,
            usesRegularExpression: true
        )
        let match = MemoTextLogic.nextMatch(in: "ITEM 12", expression: expression, startingAt: 0)
        assert(match?.range == NSRange(location: 0, length: 7))

        let literal = try MemoTextLogic.searchExpression(
            pattern: "a.b",
            usesRegularExpression: false
        )
        let wrapped = MemoTextLogic.nextMatch(in: "a.b x", expression: literal, startingAt: 5)
        assert(wrapped?.range == NSRange(location: 0, length: 3))

        assert(MemoTextLogic.isHorizontalRuleLine("___"))
        assert(MemoTextLogic.isHorizontalRuleLine("_____"))
        assert(MemoTextLogic.isHorizontalRuleLine("  ___  "))
        assert(!MemoTextLogic.isHorizontalRuleLine("__"))
        assert(!MemoTextLogic.isHorizontalRuleLine("___x"))
        assert(!MemoTextLogic.isHorizontalRuleLine(""))

        let ruleRanges = MemoTextLogic.horizontalRuleLineRanges(in: "hello\n___\nworld\n____\n" as NSString)
        assert(ruleRanges == [NSRange(location: 6, length: 3), NSRange(location: 16, length: 4)])

        assert(MemoTextLogic.horizontalRuleLineRanges(in: "no rules here" as NSString).isEmpty)
    }
}

File: Tests/MemoTextLogicCheck.cs
using Memo;

// 맥 버전 Tests/MemoTextLogicCheck.swift와 같은 항목
Check(MemoTextLogic.ListEdit("1. first", 8) == new TextEdit(8, 0, "\n2. "), "번호 목록 이어쓰기");
Check(MemoTextLogic.ListEdit("1. first\n2. ", 12) == new TextEdit(9, 3, ""), "빈 번호 항목에서 목록 끝내기");
Check(MemoTextLogic.ListEdit("- first", 7) == new TextEdit(7, 0, "\n- "), "글머리표 이어쓰기");
Check(MemoTextLogic.ListEdit("1. first\n\t- detail\n\t- ", 22) == new TextEdit(19, 3, "2. "), "하위 글머리표에서 부모 번호로 복귀");

Check(MemoTextLogic.IndentationEdit("hello", 2, 0, outdent: false) == new TextEdit(0, 5, "\thello", 3, 0), "커서 줄 들여쓰기");
Check(MemoTextLogic.IndentationEdit("one\ntwo\nthree", 1, 6, outdent: false) == new TextEdit(0, 8, "\tone\n\ttwo\n", 2, 7), "여러 줄 들여쓰기");
Check(MemoTextLogic.IndentationEdit("\tone\n\ttwo", 0, 9, outdent: true) == new TextEdit(0, 9, "one\ntwo", 0, 7), "여러 줄 내어쓰기");

var expression = MemoTextLogic.SearchRegex("item ([0-9]+)", useRegex: true);
var match = MemoTextLogic.NextMatch(expression, "ITEM 12", 0);
Check(match is { Index: 0, Length: 7 }, "정규식 대소문자 무시 검색");

var literal = MemoTextLogic.SearchRegex("a.b", useRegex: false);
var wrapped = MemoTextLogic.NextMatch(literal, "a.b x", 5);
Check(wrapped is { Index: 0, Length: 3 }, "일반 검색 처음으로 돌아가기");

Check(MemoTextLogic.IsHorizontalRuleLine("___"), "___");
Check(MemoTextLogic.IsHorizontalRuleLine("_____"), "_____");
Check(MemoTextLogic.IsHorizontalRuleLine("  ___  "), "공백 포함 ___");
Check(!MemoTextLogic.IsHorizontalRuleLine("__"), "__ 는 구분선 아님");
Check(!MemoTextLogic.IsHorizontalRuleLine("___x"), "___x 는 구분선 아님");
Check(!MemoTextLogic.IsHorizontalRuleLine(""), "빈 줄은 구분선 아님");

Check(MemoTextLogic.HorizontalRuleLineRanges("hello\n___\nworld\n____\n").SequenceEqual(new[] { (6, 3), (16, 4) }), "구분선 줄 범위");
Check(MemoTextLogic.HorizontalRuleLineRanges("no rules here").Count == 0, "구분선 없음");

Console.WriteLine("MemoTextLogic OK");

static void Check(bool condition, string name)
{
    if (!condition)
    {
        throw new Exception($"실패: {name}");
    }
}
